File: CZECHMATE/CZECHMATE/Services/FirmwareSquareNotation.swift
//
//  FirmwareSquareNotation.swift
//  Mapování „e2“ → indexy do `board[][]` (shodné s CZECHMATEShared.ChessSquareNotation): řádek 0 = rank 8, řádek 7 = rank 1.
//

import Foundation

enum FirmwareSquareNotation {
    /// Indexy do pole `snapshot.board` (stejná konvence jako `context/snapshot_golden.json`).
    static func indices(from notation: String) -> (row: Int, col: Int)? {
        let t = notation.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard t.count >= 2,
              let file = t.first,
              let rank = Int(t.dropFirst()),
              rank >= 1, rank <= 8
        else { return nil }
        let col = Int(file.asciiValue! - Character("a").asciiValue!)
        guard (0 ..< 8).contains(col) else { return nil }
        let row = 8 - rank
        guard (0 ..< 8).contains(row) else { return nil }
        return (row, col)
    }
}

File: flutter_czechmate/lib/core/models/firmware_square_notation.dart
/// Mapování „e2“ → indexy do `board[][]` (řádek 0 = rank 8, řádek 7 = rank 1).
class FirmwareSquareNotation {
  static (int row, int col)? indicesFromNotation(String notation) {
    final t = notation.trim().toLowerCase();
    if (t.length < 2) return null;
    
    final file = t.codeUnitAt(0);
    final rankChar = t.substring(1);
    final rank = int.tryParse(rankChar);
    
    if (rank == null || rank < 1 || rank > 8) return null;
    
    final col = file - 'a'.codeUnitAt(0);
    if (col < 0 || col >= 8) return null;
    
    final row = 8 - rank;
    if (row < 0 || row >= 8) return null;
    
    return (row, col);
  }
}
